README.md
# Mixcoatl
Mixcoatl was the father of Quetzalcoatl. His name means "Cloud Serpent". Fitting for a Python library for the enStratus API.

# Build Status
[![Build Status](https://secure.travis-ci.org/enStratus/mixcoatl.png)](http://travis-ci.org/enStratus/mixcoatl)

## Word of warning
This repo is not feature complete. Preference is being given to API interactions that are most useful in conjuction with Fabric:

- Managing running instances
- Managing firewalls
- Interacting with Jobs
- Support for the above (budget codes, regions, datacenters)

Also the final API is being fleshed out based on how it is being used. Currently everything except `jobs` returns a specific object type - i.e. `Server`, `Firewall`, `FirewallRule`

All objects support lazy loading as well as an `.all()` class method.

# Usage notes
At a minimum, the following environment variables will need to be set:

- `ES_SECRET_KEY`
- `ES_ACCESS_KEY`

By default, api calls will be made against the enStratus production SaaS and API version `2012-06-15`. These can be overridden with the following variables:

- `ES_ENDPOINT`
- `ES_API_VERSION`

When overriding the endpoint, it should be in the form of:

`http[s]://api.endpoint.domain/enstratus/api/<api version>`

Note that setting both `ES_API_VERSION` and `ES_ENDPOINT` is not cumulative. If you wish to use a private endpoint, it must include the version in the url.

## Lazy loading
By default, any object you request by its id will not actually hit the endpoint. Only when you request the object in full or a specific attribute, will it actually make the API call. If the API call fails, the error will be returned to you. You can always check the object's `last_error` attribute to determine if it failed or not.

example:

```python
from mixcoatl.geography.cloud import Cloud
c = Cloud(1)
>>> # returns immediately
c.cloud_id
>>> 1
c.name
>>> # api call is made.
>>> u'Amazon Web Services'
```

## `.all()`
All objects should support a call to return all resources of that type. This will actually return a list of objects. Note that calling `.all()` actually deferences the objects so an API call will be made for each object:

example:

```python
from mixcoatl.geography.cloud import Cloud
c = Cloud.all()
>>> # Initial call made for all Clouds
c
>>> # Delay while each cloud object is dereferenced
c[0]
>>> {'status': 'ACTIVE', 'current_job': None, 'last_request': '<Response [200]>', 'name': 'Amazon Web Services', 'last_error': None, 'cloud_provider_name': 'Amazon', 'cloud_provider_console_url': 'http://aws.amazon.com', 'cloud_provider_logo_url': '/clouds/aws.gif', 'compute_endpoint': 'https://ec2.us-east-1.amazonaws.com,https://ec2.us-west-1.amazonaws.com,https://ec2.eu-west-1.amazonaws.com', 'compute_secret_key_label': 'AWS_SECRET_ACCESS_KEY', 'documentation_label': None, 'compute_delegate': 'org.dasein.cloud.aws.AWSCloud', 'path': 'geography/Cloud/1', 'compute_account_number_label': 'AWS_ACCOUNT_NUMBER', 'private_cloud': False}
type(c[0])
>>> mixcoatl.geography.cloud.Cloud
c[0].__class__.__name__
>>> 'Cloud'
```

## Pretty-printing
Every resource has a `.pprint()` function available which returns the prettyprinted object

example:
```python
from mixcoatl.geography.cloud import Cloud
c = Cloud(1)
c
c.pprint()
>>> # pretty print representation
```

# Other notes
In general, most resources should support read-only access. If you know the id of an resource, you can simply request the resource by name with the id as a parameter:

```python
from mixcoatl.infrastructure.server import Server
s = Server(12345)
s
>>> # server details returned
```

Importing resources generally follows the API directly i.e.:

`from mixcoatl.scope.resource import ResourceName`

so for firewalls:

```python
from mixcoatl.network.firewall import Firewall
f = Firewall(12345)
f
```

for servers:

```python
from mixcoatl.infrastucture.server import Server
s = Server(12345)
s
```

For specific examples per resource, see the [wiki](https://github.com/enStratus/mixcoatl/wiki)


CHANGELOG.md
Mixcoatl ChangeLog
==================

![Mixcoatl Snake](http://mixcoatl.net/assets/images/mixcoatl_serpent.png)

Unreleased
===========
Placeholder for changes pending in the next release

Release Date: pending

You can now configure multiple alternate DCM endpoint via a new python object Endpoint.

Features
---------
- Added [#225][225]: multiple endpoint support

Fixes
--------
- Fixed [#207][207]: allow configuration via python

[207]:https://github.com/enStratus/mixcoatl/issues/207
[225]:https://github.com/enStratus/mixcoatl/pull/225


0.10.50.2
=============

Release Date: 2015-07-16

Fixes
------
- Fixed [#234][234]: fix missing import statement in dcm-list-datacenters. Thanks Jim Sander for the 
  report.

[234]:https://github.com/enStratus/mixcoatl/issues/234

0.10.50.1
=============

Release Date: 2015-07-10

Fixes
------
- Fixed [#216][216]: add network and subnet to the server class and dcm-create-server cli. Thanks to
  Eoin Kennet at HEAnet for the bug report.

[216]:https://github.com/enStratus/mixcoatl/issues/216

0.10.50
=============

Release Date: 2015-07-07

Features
--------
- Added [#202][202]: dcm-create-user support for ssh key
- Added [#214][214]: handsfree installer support for ldap dirsync, saml config, private cloud pricing, and marketplace setup
- Added [#229][229]: runtime adding of access methods for new API attributes

Fixes
------
- Fixed [#217][217] [#221][221]:  some dcm-* cli tools crash when pretty printing missing attributes
- Fixed [#226][226]: dcm-list-volumes not reporting server ID
- Fixed [#212][212]: default DCM_ENDPOINT for dcm.enstratius.com

[202]:https://github.com/enStratus/mixcoatl/pull/202
[214]:https://github.com/enStratus/mixcoatl/pull/214
[229]:https://github.com/enStratus/mixcoatl/issues/229
[217]:https://github.com/enStratus/mixcoatl/pull/217
[221]:https://github.com/enStratus/mixcoatl/issues/221
[226]:https://github.com/enStratus/mixcoatl/issues/226
[229]:https://github.com/enStratus/mixcoatl/issues/229
[212]:https://github.com/enStratus/mixcoatl/pull/212







mixcoatl/resource.py
"""
mixcoatl.resource
------------------
"""
import os
import requests as r
import mixcoatl.auth as auth
from mixcoatl.settings.load_settings import settings
from mixcoatl.decorators.lazy import lazy_property
from mixcoatl.utils import camel_keys
import json
import datetime


class Endpoint(object):

    """ A class for representing DCM endpoints and loading them from json files. Each Endpoint contains not only the URL
    for the endpoint but also the associated API keys.

    :param url: The url of the DCM endpoint (e.g https://dcm.enstratius.com/api/enstratus/2015-05-25 )
    :param basepath: Optional server path to the API (e.g. /api/enstratus/2015-05-25)
    :param api_version: required API version (e.g. 2015-05-25) , if missing an attempt is made to parse it
                        from from the end of the url
    :param access_key: DCM access key
    :param secret_key: DCM secret key
    :param ssl_verify: Should the endpoints SSL key be verified. Default is True
    :param nickname: optional arbitrary short nick name to assign this endpoint

    Here's an example of how you could use endpoints to identify servers simultaneously managed by two DCM instances:

        .. code-block:: python

            from mixcoatl.infrastructure.server import Server
            from mixcoatl.resource import Endpoint

            saas_endpoint = Endpoint(nickname="saas",
                                     url="https://dcm.enstratius.com/api/enstratus/2015-05-25",
                                     api_version="2015-05-25",
                                     access_key="POIUYTREQ",
                                     secret_key="ukjdf8HydvkLlki=4hfksj")

            vagrant_endpoint = Endpoint(nickname="vagrant",
                                        url="https://vagrant.vm/api/enstratus/2015-05-25",
                                        api_version="2015-05-25",
                                        access_key="LKJHGFDSAQ",
                                        secret_key="ht748fbsd974d=t874gDFb",
                                        ssl_verify=False)


            saas_servers = Server.all(endpoint=saas_endpoint)
            vagrant_servers = Server.all(endpoint=vagrant_endpoint)

            # find all the servers managed by two DCM instances
            for s in saas_servers:
                for v in vagrant_servers:
                    if s.provider_id == v.provider_id:
                        print "SaaS: %s Vagrant: %s, Provider ID %s" % (s.server_id, v.server_id, v.provider_id)




    """

    def __init__(self, url=None, basepath=None, api_version=None, secret_key=None,
                 access_key=None, ssl_verify=True, nickname=None):
        """

        """
        if url:
            self.url = url
        else:
            raise ValueError("endpoint url must be specified")
        if api_version:
            self.api_version = self._validate_api_version(api_version)
        else:
            try:
                # attempt to grab the API version from the url
                self.api_version = self._validate_api_version(self.url.split('/')[-1])
            except ValueError:
                raise ValueError("endpoint was created without an api_version and did not contain a " +
                                 "URL terminated with a valid date in YYYY-MM-DD format")
        if basepath:
            self.basepath = basepath
        else:
            self.basepath = "/api/enstratus/%s" % self.api_version
        if access_key:
            self.access_key = access_key
        else:
            raise ValueError("access_key must be specified")
        if secret_key:
            self.secret_key = secret_key
        else:
            raise ValueError("secret_key must be specified")

        self.ssl_verify = ssl_verify
        self.nickname = nickname

    @staticmethod
    def _validate_api_version(date):
        try:
            datetime.datetime.strptime(date, '%Y-%m-%d')
            return date
        except ValueError:
            raise ValueError("Incorrectly formatted API version, should be YYYY-MM-DD")

    @classmethod
    def from_file(cls, filename):
        """ Load a DCM endpoint definition from a json file and return and Endpoint object.

        :param str filename: the full path to a json file containing and endpoint definition
        :returns: the endpoint loaded from the file
        :rtype: Endpoint

        An example json file containing the containing the endpoint definition:

            .. code-block:: json
            
                {
                  "nickname":"saas",
                  "url":"https://dcm.enstratius.com/api/enstratus/2015-05-25",
                  "api_version":"2015-05-25",
                  "access_key":"abcdefg",
                  "secret_key":"gfedcba",
                  "ssl_verify": true
                }
        """

        endpoint_json = open(filename).read()
        e = json.loads(endpoint_json)

        return Endpoint(url=e['url'],
                        basepath=e['basepath'] if 'basepath' in e else None,
                        api_version=e['api_version'],
                        secret_key=str(e['secret_key']),
                        access_key=str(e['access_key']),
                        ssl_verify=e['ssl_verify'] if 'ssl_verify' in e else None,
                        nickname=e['nickname'] if 'nickname' in e else None)

    @classmethod
    def multiple_from_file(cls, filename):
        """ Load multiple DCM endpoint definitions from a json file and return a dictionary indexed by Endpoint
         nicknames.

        :param str filename: the full path to a json file containing and multiple endpoint definitions
        :returns: a dictionary of Endpoints indexed by endpoint nicknames
        :rtype: dictionary of Endpoint objects

        Here is an example json file that can be loaded by this function:

            .. code-block:: json

                [
                  {
                    "nickname": "saas",
                    "url": "https://dcm.enstratius.com/api/enstratus/2015-05-25",
                    "api_version": "2015-05-25",
                    "access_key": "abcdefg",
                    "secret_key": "gfedcba",
                    "ssl_verify": true
                  },
                  {
                    "nickname": "vagrant",
                    "url": "https://vagrant.vm/api/enstratus/2015-05-25",
                    "api_version": "2015-05-25",
                    "access_key": "abcdefg",
                    "secret_key": "gfedcba",
                    "ssl_verify": true
                  }
                ]
        """

        endpoint_json = open(filename).read()
        endpoint_in = json.loads(endpoint_json)
        endpoint_dict = {}

        for e in endpoint_in:
            endpoint_dict[e['nickname']] = Endpoint(url=e['url'],
                                                    basepath=e['basepath'] if 'basepath' in e else None,
                                                    api_version=e['api_version'],
                                                    secret_key=str(e['secret_key']),
                                                    access_key=str(e['access_key']),
                                                    ssl_verify=e['ssl_verify'] if 'ssl_verify' in e else None,
                                                    nickname=e['nickname'] if 'nickname' in e else None)
        return endpoint_dict


class Resource(object):

    """The base class for all resources returned from an DCM API call
    By default all resources are largely represented as a `dict`-alike object
    that mirrors the JSON response from the DCM API with keys converted
    from camel-case to snake-case.

    For instance:

        * The DCM API JSON for a resource looks something like this:

            .. code-block:: yaml

                {
                  'someId':1,
                  'name':'my-name',
                  'cloud':{'cloudId':12345}
                }

        * All keys (nested or top-level) will be converted to snake-case:
            - `someId` becomes `some_id`
            - `cloudId` becomes `cloud_id`
        * The top-level keys will be converted to getters (and in some cases - setters) on the Resource.
            - `some_id` will now be the resource's primary identifier: `resource.some_id`
            - `name` will likely be a getter and a setter as `name` is usually a mutable attributes
            - `cloud` will be converted to a setter with a value of `{'cloud_id':123345}`
        * In cases where a setter would actually need to set a nested value, it will do so. However
            the getter will still return the original data structure as appropriate

        .. warning::

            The Resource object, while looking much like a `dict`, is not a `dict` proper.
            If you need an actual `dict`, call `to_dict()` on your instance.
    """

    #: The base request path of the resource
    PATH = None
    #: The top-level grouping of a resource in the API response
    COLLECTION_NAME = None
    #: The unique identifier of an individual resource
    PRIMARY_KEY = None

    def __init__(self, base_path=None, request_details='basic', endpoint=None, **kwargs):
        if base_path is None:
            try:
                self.__path = self.__class__.PATH
            except:
                raise AttributeError('you must override base_path')
        else:
            self.__path = base_path

        if 'request_details' in kwargs:
            self.__request_details = kwargs['request_details']
            del kwargs['request_details']
        else:
            self.__request_details = 'basic'

        if 'params' in kwargs:
            self.__params = kwargs['params']
        else:
            self.__params = {}
        if endpoint:
            self.__endpoint = endpoint
        else:
            # no endpoint provided, pull from settings
            self.__endpoint = Endpoint(url=settings.endpoint,
                                       basepath=settings.basepath,
                                       api_version=settings.api_version,
                                       secret_key=settings.secret_key,
                                       access_key=settings.access_key,
                                       ssl_verify=settings.ssl_verify)
        self.__last_request = None
        self.__last_error = None
        self.__current_job = None
        self.__status_code = None
        self.__payload_format = 'json'
        self.__request_details = request_details
        self.pending_changes = {}

    def __props(self):
        """List of properties and lazy properties for a given resource"""
        p = [k for k, v in self.__class__.__dict__.items() if type(
            v) in [lazy_property, property]]
        rp = ['last_error', 'path', 'last_request',
              'current_job', 'request_details']
        return p + rp

    def __repr__(self):
        from mixcoatl.utils import convert
        d = {}
        for x in self.__props():
            try:
                if x == 'last_request':
                    d[x] = str(getattr(self, x))
                else:
                    d[x] = getattr(self, x)
            except AttributeError:
                d[x] = None
            except KeyError:
                d[x] = None
        return repr(convert(d))

    @property
    def endpoint(self):
        return self.__endpoint

    @property
    def status_code(self):
        return self.__status_code

    @property
    def request_details(self):
        """The level of detail used in the API call: `basic` or `extended`"""
        return self.__request_details

    @request_details.setter
    def request_details(self, level):
        self.__request_details = level

    @property
    def payload_format(self):
        """The format of the payload: `json` or `xml`"""
        return self.__payload_format

    @payload_format.setter
    def payload_format(self, p_format):
        self.__payload_format = p_format

    @property
    def path(self):
        """The path used in the API request (e.g. ``admin/Job``)"""
        return self.__path

    @path.setter
    def path(self, p):
        self.__path = p

    @property
    def last_request(self):
        """The :class:`Request` object of the most recent API call"""
        return self.__last_request

    @last_request.setter
    def last_request(self, lr):
        self.__last_request = lr

    @property
    def last_error(self):
        """The last error message, if any, returned from the most recent API call"""
        return self.__last_error

    @last_error.setter
    def last_error(self, le):
        self.__last_error = le

    @property
    def current_job(self):
        """The current :class:`Job`, if any, of an asynchronous API call"""
        return self.__current_job

    @current_job.setter
    def current_job(self, cj):
        self.__current_job = cj

    @property
    def params(self):
        return self.__params

    @params.setter
    def params(self, p):
        self.__params = p

    def add_property(self, prop_name):
        """Adds setter, getter and property for field `prop_name`"""
        def set_attribute(cls, val):
            # since we are using __ mangling we need _'+cls.__class__.__name__+
            setattr(cls, '_'+cls.__class__.__name__+'__'+prop_name, val)

        def get_attribute(cls):
            return getattr(cls, '_'+cls.__class__.__name__+'__'+prop_name)

        setattr(self.__class__, 'set_'+prop_name, set_attribute)
        setattr(self.__class__, 'get_'+prop_name, get_attribute)
        setattr(self.__class__,
                prop_name,
                property(
                    getattr(self.__class__, 'get_'+prop_name),
                    getattr(self.__class__, 'set_'+prop_name)
                ))

    def load(self, **kwargs):
        """(Re)load the current object's attributes from an API call"""
        from mixcoatl.utils import uncamel_keys
        reserved_words = ['type']
        p = self.PATH + "/" + str(getattr(self, self.__class__.PRIMARY_KEY))

        if 'params' in kwargs:
            params = kwargs['params']
        else:
            params = camel_keys(self.params)

        s = self.get(p, params=params)
        if self.last_error is None:
            scope = uncamel_keys(s[self.__class__.COLLECTION_NAME][0])
            for k in scope.keys():
                if k in reserved_words:
                    the_key = 'e_' + k
                else:
                    the_key = k
                nk = '_%s__%s' % (self.__class__.__name__, the_key)
                if the_key not in self.__props():
                    self.add_property(k)
                    if 'DCM_DEBUG' in os.environ:
                        print "Missing key found when loading class %s with primary key %s: %s" % \
                              ( self.__class__.__name__, str(getattr(self, self.__class__.PRIMARY_KEY)), k)
                setattr(self, nk, scope[k])
                self.loaded = True
        else:
            return self.last_error

    def __doreq(self, method, **kwargs):
        """Performs the actual API call

        * calls `auth.get_sig` for signed headers
        * issues the requested :attr:`method` against the API endpoint
        * Handles requests appropriately based on sync/async nature of the call
            based on DCM API documentation
        """
        failures = [400, 403, 404, 409, 500, 501, 503]
        sig = auth.get_sig(method, self.path, endpoint=self.__endpoint)
        url = self.__endpoint.url + '/' + self.path
        ssl_verify = self.__endpoint.ssl_verify

        if self.payload_format == 'xml':
            payload_format = 'application/xml'
        elif self.payload_format == 'json':
            payload_format = 'application/json'
        else:
            raise AttributeError(
                'Wrong payload format: %s' % self.payload_format)

        headers = {'x-esauth-access': sig['access_key'],
                   'x-esauth-timestamp': str(sig['timestamp']),
                   'x-esauth-signature': str(sig['signature']),
                   'x-es-details': self.request_details,
                   'Accept': payload_format,
                   'User-Agent': sig['ua']}

        results = r.request(
            method, url, headers=headers, verify=ssl_verify, **kwargs)

        if 'DCM_DEBUG' in os.environ:
            print "URL: %s" % (url)
            print "method: %s" % (method)
            for key, value in kwargs.iteritems():
                print "%s = %s" % (key, value)
            for key, value in results.headers.iteritems():
                print "headers: %s = %s" % (key, value)

        self.last_error = None
        self.last_request = results
        self.__status_code = results.status_code

        if self.payload_format == 'xml':
            return results

        if results.status_code in failures:
            try:
                err = results.json()
                self.last_error = err['error']['message']
            except ValueError:
                self.last_error = results.content
            return self.last_error

        if method == 'GET':
            try:
                results.raise_for_status()
                self.last_error = None
                return results.json()
            except r.exceptions.HTTPError:
                try:
                    err = results.json()
                    self.last_error = err['error']['message']
                except ValueError:
                    self.last_error = results.content
                return False
        if method == 'DELETE':
            if results.status_code == 202:
                self.current_job = results.json()['jobs'][0]['jobId']
                return results.json()
            elif results.status_code != 204:
                try:
                    err = results.json()
                    self.last_error = err['error']['message']
                except ValueError:
                    self.last_error = results.content
                return False
            else:
                return True
        if method == 'PUT':
            if results.status_code == 202:
                self.current_job = results.json()['jobs'][0]['jobId']
                return results.json()
            elif results.status_code == 204:
                return True
            else:
                try:
                    err = results.json()
                    self.last_error = err['error']['message']
                except ValueError:
                    self.last_error = results.content
                return False
        if method == 'POST':
            if results.status_code in [201, 202]:
                if results.status_code == 202:
                    self.current_job = results.json()['jobs'][0]['jobId']
                return results.json()
            else:
                try:
                    err = results.json()
                    self.last_error = err['error']['message']
                except ValueError:
                    self.last_error = results.content
                return False
        if method == 'HEAD':
            if results.status_code in [200]:
                try:
                    err = results.json()
                    self.last_error = err['error']['message']
                except ValueError:
                    self.last_error = results.content
                return False

    def set_path(self, path=None):
        if path is not None:
            self.path = path

    def get(self, path=None, **kwargs):
        """Perform an HTTP `GET` against the API endpoint for the current resource"""
        self.set_path(path)
        return self.__doreq('GET', **kwargs)

    def post(self, path=None, **kwargs):
        """Perform an HTTP `POST` against the API endpoint for the current resource"""
        self.set_path(path)
        return self.__doreq('POST', **kwargs)

    def put(self, path=None, **kwargs):
        """Perform an HTTP `PUT` against the API endpoint for the current resource"""
        self.set_path(path)
        return self.__doreq('PUT', **kwargs)

    def head(self, path=None, **kwargs):
        """Perform an HTTP `HEAD` against the API endpoint for the current resource"""
        self.set_path(path)
        return self.__doreq('HEAD', **kwargs)

    def delete(self, path=None, **kwargs):
        """Perform an HTTP `DELETE` against the API endpoint for the current resource"""
        self.set_path(path)
        return self.__doreq('DELETE', **kwargs)

    def pprint(self):
        """The prettyprint formatted representation of the current resource"""
        import pprint
        pprint.pprint(eval(repr(self)))

    def to_dict(self):
        """The `dict` representation of the current resource"""
        return eval(repr(self))

    def track_change(self, var, prev, new):
        if prev == new:
            pass
        else:
            self.pending_changes[var] = {'old': prev, 'new': new}
